python/ray/worker.py

```diff
@@ -1231,6 +1236,10 @@ def connect(node,
     # will use glog, which is intialized in `CoreWorker`.
     ray.state.state._initialize_global_state(
         node.redis_address, redis_password=node.redis_password)
+    worker_job_config = worker.core_worker.get_job_config()
+    package_uri = worker_job_config.runtime_env.working_dir_uri
+    if mode == SCRIPT_MODE or mode == WORKER_MODE:
```

Comment:
Can we only call this for workers?

Reply:
No. driver and worker do different things.
driver is to upload and worker is to download.

Reply:
Can you move the driver package upload code to init()?

Reply:
I need to do these things:

1. Job starts (but, don't prewarm workers yet)
2. Put to storage any missing data (associate with job id)
3. Finishes: can start job workers now)

1 and 3 are in connect, so 2 has to be in connect if I understand it correctly.

Any reason you want to move it to init?

Reply:
I see, so the reason it has to be here is because the job id isn't available until connect, is that right? How about this then...
```
if mode == SCRIPT_MODE:
   # Upload the package and associate it with the job. We can assume the package
   # was created locally already in init().
   ray_pkg.upload_for_job(package_uri, job_id)
elif mode == WORKER_MODE:
   # Download and setup the working dir if it doesn't exist.
   ray_pkg.download_and_setup(package_uri)
```

Reply:
Not because of job id, but because uploading needs `core_work` to be created.